### PR TITLE
 planner/core: support query block hint

### PR DESCRIPTION
**Pull request by lamxTyler**
### What problem does this PR solve? 

Support using query block in optimizer hint.

### What is changed and how it works?

This PR will first tranverse the ast and assign the offset of stmt within the sql from left to right, then group hints in a top down style and let them take effects at specific query block.

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change

Side effects

 - None

Related changes

 - None

### CODE REVIEW

File: planner/core/hints.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+		}
+	}
+	if qbName == "" {
+		return
+	}
+	if p.QbNameMap == nil {
+		p.QbNameMap = make(map[string]int)
+	}
+	if _, ok := p.QbNameMap[qbName]; ok {
+		p.Ctx.GetSessionVars().StmtCtx.AppendWarning(errors.New(fmt.Sprintf("Duplicate query block name %s, we will use the first one", qbName)))
```
Comment (eurekaka) [inline planner/core/hints.go]:
```suggestion
		p.Ctx.GetSessionVars().StmtCtx.AppendWarning(errors.New(fmt.Sprintf("Duplicate query block name %s, only the first one is effective", qbName)))
```

File: planner/core/hints.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+// Enter implements Visitor interface.
+func (p *BlockHintProcessor) Enter(in ast.Node) (ast.Node, bool) {
+	switch node := in.(type) {
+	case *ast.UpdateStmt:
+		p.checkQueryBlockHints(node.TableHints, 0)
+	case *ast.DeleteStmt:
+		p.checkQueryBlockHints(node.TableHints, 0)
+	case *ast.SelectStmt:
+		p.selectStmtOffset++
+		node.QueryBlockOffset = p.selectStmtOffset
```
Comment (eurekaka) [inline planner/core/hints.go]: Shall we start the naming from `sel_0` or `sel_1`?

File: planner/core/hints.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	if p.QbHints == nil {
+		p.QbHints = make(map[int][]*ast.TableOptimizerHint)
+	}
+	for _, hint := range hints {
+		if hint.HintName.L == "qb_name" {
+			continue
+		}
+		offset := p.getHintOffset(hint, nodeType, currentOffset)
+		if offset < 0 {
+			continue
```
Comment (eurekaka) [inline planner/core/hints.go]: Shall we append a warning for this case?

File: planner/core/hints.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+		if topOffset < 0 {
+			return -1
+		}
+	}
+	if len(hint.Tables) == 0 {
+		return topOffset
+	}
+	// Handle the case of hint_name(t1@sel_1, t2@sel_2), the result offset should be the out most stmt, which has the
+	// smallest block offset.
+	minOffset := math.MaxInt64
```
Comment (eurekaka) [inline planner/core/hints.go]: Could you give a full example of this case?

File: planner/core/hints.go (hunk 2 above)
Comment (lamxTyler) [reply planner/core/hints.go]: `sel_1`, always number from 1.

File: planner/core/hints.go (hunk 3 above)
Comment (lamxTyler) [reply planner/core/hints.go]: It is already explained in the comment, what infomation do you still need?

File: planner/core/hints.go (hunk 4 above)
Comment (lamxTyler) [reply planner/core/hints.go]: It is already explained in the comment, what kind of infomation do you still need?

Comment (eurekaka) [reply planner/core/hints.go]: How to understand `hint_name(t1@sel_1, t2@sel_2)`? How is this kind of hint being used? Is it for join order hint?

File: planner/core/hints.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+// Leave implements Visitor interface.
+func (p *BlockHintProcessor) Leave(in ast.Node) (ast.Node, bool) {
+	return in, true
+}
+
+// checkQueryBlockHints checks the validity of query blocks and records the map of query block name to select offset.
+func (p *BlockHintProcessor) checkQueryBlockHints(hints []*ast.TableOptimizerHint, offset int) {
+	var qbName string
+	for _, hint := range hints {
+		if hint.HintName.L != "qb_name" {
```
Comment (winoros) [inline planner/core/hints.go]: make `qb_name` as a const?

File: planner/core/hints.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+// getHintOffset gets the offset of stmt that the hints take effects.
+func (p *BlockHintProcessor) getHintOffset(hint *ast.TableOptimizerHint, nodeType nodeType, currentOffset int) int {
+	if hint.QBName.L != "" {
+		return p.getBlockOffset(hint.QBName, nodeType)
+	}
+	return currentOffset
+}
+
+func (p *BlockHintProcessor) getCurrentStmtHints(hints []*ast.TableOptimizerHint, nodeType nodeType, currentOffset int) []*ast.TableOptimizerHint {
```
Comment (winoros) [inline planner/core/hints.go]: add some header comments?

**Pull request merged**